### PR TITLE
Fixes to general sorting algorithms for hoomd and lammps

### PR DESCRIPTION
**Pull request by CalCraven**
HOOMD functionality right now leads to bugs in writing out the forcefields.

There are two functions that need to be in sync
`gmso/external/convert_hoomd.py to_hoomd_snapshot`
`gmso/external/convert_hoomd.py to_hoomd_forcefield`

These are not compatible, as discussed in #758.

Additionally, the sorting algorithms in gmso.utils.sorting.py have some bugs with dihedrals, which need to be flipped if the two central atoms are equivalent, but the first and fourth atoms should be in reversed order.

This PR will restructure some of the sorting algorithms and migrate some of the ones that were used in formats/lammpsdata.py from gmso/core/views.py to gmso/utils/sorting.py

TODO List
- [x] unit tests
- [x] Move all sorting functions into
- [x] Fixes to sort_connection_members
- [x]  Fix HOOMD to sort by site.atom_type.atomclasses
- [x] Work with wildcards
- [ ] issue raised and addressed

### CODE REVIEW

**Comment by CalCraven** [comment]
> One high-level comment I have (probably need a broader discussion later too), so it seems right now, we are going with atom class for everything (storing bond and bond params in the snapshot and forcefield), which would work for most of our current use case, i.e., when connection types are defined between atom classes. This, however, may throw an issue if there is a connection that is defined between atom types. The latter case is rarer but, IMO, it might happen and should be supported. In other word, I think we should go with atom types for everything instead of atomclass. 

So I think it's a necessity to use atomclass instead of atomtype. This is related to #765, which is to try to get hoomd to support GAFF. In fact, that PR is updated with the changes to this PR, because both are necessary for GAFF, so in the end I'll probably address these comments/fixes directly in there and close this PR. But I'm getting away from the point. Here's the situation that arises.

A topology has 5 atomtypes, each with unique names, bonded in a line to form 2 unique dihedrals. When writing out the snapshot and the forcefield, the important attribute is the string that is used as the key for these lookup tables. If we use atomtypes, then it'll look like:

Molecule representation -> atom1 - atom2 - atom3 - atom4 - atom5
In snapshots
Dihedral1 = "atype1-atype2-atype3-atype4"
Dihedral2 = "atype2-atype3-atype4-atype5"

in forces
Dihedral_type1 = "atype1-atype2-atype3-atype4"
Dihedral_type2 = "atype2-atype3-atype4-atype5"

But if the classes of dihedrals are all the same, then we're duplicating dihedrals.
Dihedral_type1 = Dihedral_type2. If we used classes, we would reduce the number of dihedrals written. And 90% or more of our forcefields are defined by atomclasses, and with big forcefields we're at risk of writing huge numbers of dihedrals out with all of the duplication.

Now the situation you're describing, where the dihedrals are defined by their atomtypes, not atomclasses. In this scenario, the atomtypes = atomclasses.  
atomtype1 = "OPLS_135" atomclass1 = "OPLS_135". In that case, we should still be okay to use the classes as they are equivalent in using their strings. 

The strange situation would be: 
atomtype1 = "OPLS_135" atomclass1 = "CT"
atomtype2="OPLS_136 atomclass1="CT" 
but two different dihedrals are defined as 
"OPLS_135-OPLS_136-X-X"
"CT-CT-X-X"
This would write these two dihedrals the same way. But I don't thing writing the atomtypes would be a useful solution here as well. Because then what are the atomtypes for the "CT-CT-X-X" dihedral? In theory it could be any combination of atomtypes that correlate to that class. So I'm not sure we could handle that situation either way, unless we know the original formulation of the dihedral.
```python
for dihedral in top.dihedrals:
        unique_members = tuple(
            [site.atom_type.atomclass for site in dihedral.connection_members]
        )
        unique_dihedrals[unique_members] = dihedral
```

File: gmso/utils/sorting.py [hunk 1]
```diff
@@ -63,13 +53,109 @@ def sorting_key(site):
...
+            return [site1, site2, site3, site4]
     elif isinstance(connection, gmso.Improper):
         site1, site2, site3, site4 = connection.connection_members
         site2, site3, site4 = sorted([site2, site3, site4], key=sorting_key)
         return [site1, site2, site3, site4]
     else:
         raise TypeError("Provided connection not supported.")
+
+
+def sort_by_classes(potential):
```
[the remark this answers is not in the document]
Comment (CalCraven) [reply gmso/utils/sorting.py:71]: Fixed by adding more detailed doc strings. I don't think every function needs to be that detailed. `potential` is the only argument, so we don't need it to also be part of the function name. I added details about what potential is and how it's sorted.

File: gmso/utils/sorting.py [hunk 2]
```diff
@@ -14,42 +32,14 @@ def natural_sort(text):
...
-    elif isinstance(connection_type, gmso.ImproperType):
-        type1, type2, type3, type4 = connection_type.member_types
-        type2, type3, type4 = sorted([type2, type3, type4], key=natural_sort)
-        return [type1, type2, type3, type4]
-    else:
-        raise TypeError("Provided connection_type not supported.")
-
-
 def sort_connection_members(connection, sort_by="name"):
     """Sort connection_members of connection."""
```
[the remark this answers is not in the document]
Comment (CalCraven) [reply gmso/utils/sorting.py:36]: done.

**Comment by daico007** [comment]
I am not too worry about the duplication issue (since that will be more or less constrained by how many unique molecules in a systems and how many unique dihedrals in those molecules). There will be duplications, but I don't think that will be to that much of a degree. Consider that against the trade-off of supporting the 10% non-common case, that might be worth it. 

What I mean about dihedral define between atomtypes, I mean it's look like:
`type1="aa" type2="bb" type3="cc" type4="dd"`. Making atomclass to take the value of atomtype is just a work-around.

This topic has been brought up a while ago, but in the strange case that you brought up, the dihedral define by type will take precedent to that defined by atomclass.

**Comment by CalCraven** [comment]
I think the duplication issue will get out of hand. For instance, I have simulations with up to 20 different atomtypes in them. But maybe only 10 or so atomclasses. I have the feeling that will result in a large number of dihedrals between the two cases. There's a reason something large like the OPLS forcefield defines the dihedrals by classes, not by types. I will put together an example of this and we can count the number of dihedrals, but run time will be affected by the number of dihedrals in hoomd no doubt about it. I've never actually seen a forcefield with hybrid type and class definitions. It seems a bit silly: if you have a unique site that needs it's own unique dihedral, then give it an atomclass that's the same as its atomtype. We certainly do not have any tested examples of such a forcefield if it is something we would want to support anyways.

**Comment by CalCraven** [comment]
I also am totally spitballing with that 10% uncommon case. I've never actually seen that done. So who knows exactly what that number is.

**Comment by CalCraven** [comment]
if the dihedral is type1="aa" and is given no class1, then it's class should also be "aa".

**Comment by daico007** [comment]
> It seems a bit silly: if you have a unique site that needs it's own unique dihedral, then give it an atomclass that's the same as its atomtype. 

I am thinking more about the case where the bond/angle is defined on atomclass and the dihedral on atomtype (for some reason, like you try to develop a forcefield I try to have it very specific to the atom type level). In that case, the atom type still need its original atomclass info.

**Comment by CalCraven** [comment]
Having just checked [this](https://www.watson-int.com/alkyl-polyglucoside-apg-cas-68515-73-1/), it resulted in 23 unique dihedrals by types and 14 by classes.

**Comment by CalCraven** [comment]
> I am thinking more about the case where the bond/angle is defined on atomclass and the dihedral on atomtype (for some reason, like you try to develop a forcefield I try to have it very specific to the atom type level). In that case, the atom type still need its original atomclass info.

I can see this. We need tests for it though, which we don't currently have. I think we could have a check at the beginning of the writer that checks to see if all the classes capture all of the unique dihedral expressions. If not, then we populate the snapshot and forces by types, but that shouldn't be the default case.

I think this PR should get merged tomorrow for CECAM, and I don't have the time to add that feature in. So potentially we can just raise this as a bug in the issues, and address it when we have the time/ a reasonable test case.

**Comment by daico007** [comment]
included in #765

**Pull request closed without merging**